Meta: Clearly document current guidance for changing existing PEPs (#2378)
python/peps: CONTRIBUTING.rst

developing at least a prototype reference implementation of your idea.

Contributing changes to existing PEPs
-------------------------------------

In general, most non-Draft/Active PEPs are considered to be historical
documents rather than living specifications or documentation. Major changes to
their core content usually require a new PEP, while smaller modifications may
or may not be appropriate, depending on the PEP's status. See `PEP 1

> <https://www.python.org/dev/peps/pep-0001/#pep-maintenance>`__ for more.

Reviewer: I understand this is just a summary, but there's such a big divide between Draft and Active PEPs (SC approval comes to mind :-) that it's wrong to describe them as more or less the change here.

Reviewer: (Sorry, I confused Active with Accepted. Never mind then.)

Copyediting and proofreading Draft and Active PEPs (including fixing spelling

> and formatting issues) is welcomed, and can be done via pull request to this

Reviewer: I think the author should be allowed to have total control. In the end the PEP repo is a convenient place where authors place their proposals to be discussed by the community. Authors may appreciate help, but PEPs, unlike general documentation, *have* authors, and the process is geared towards their needs.

Author: I originally included an explicit conditional on this line, "at the discretion of the author(s)", but eliminated it in response to the request to make the more concise. I added back appropriate caveats both places this is referred to.

(As a sidenote, there are _some_ limits to what is an author's discretion over a PEP editor, when push comes to shove—insensitive or inappropriate language, syntax/technical issues, major breaches of PEP 1/12, etc. But not need to delve into that here.)

developing at least a prototype reference implementation of your idea.

Contributing changes to existing PEPs
-------------------------------------

In general, most non-Draft/Active PEPs are considered to be historical
documents rather than living specifications or documentation. Major changes to
their core content usually require a new PEP, while smaller modifications may
or may not be appropriate, depending on the PEP's status. See `PEP 1
<https://www.python.org/dev/peps/pep-0001/#pep-maintenance>`__ for more.

Copyediting and proofreading Draft and Active PEPs (including fixing spelling
and formatting issues) is welcomed (subject to review by the PEP authors),
and can be done via pull request to this repo, though we generally advise
against PRs that simply mass-correct minor typos on older PEPs which don't
significantly impair meaning and understanding.
Substantive content changes should first be proposed on PEP discussion threads.

If you're still unsure, we encourage you to reach out first before opening a
PR here. For example, you could contact the PEP author(s), propose your idea in

> a discussion venue appropriate to the PEP (such as `Typing-SIG
> <https://mail.python.org/archives/list/typing-sig@python.org/>`__ for static
> typing, or `Packaging Discourse <https://discuss.python.org/c/packaging/>`__
> for packaging), or `open an issue <https://github.com/python/peps/issues>`__.
> 

Author: Your suggestion with fixed line breaks:

```suggestion
a discussion venue appropriate to the PEP's topic, or
`open an issue here <https://github.com/python/peps/issues>`__.
```

Commit messages and PR titles
-----------------------------

When adding or modifying a PEP, please include the PEP number in the commit
summary and pull request title. For example, ``PEP NNN: <summary of changes>``.

> Likewise, prefix rendering infrastructure changes with ``Infra:``, linting
> alterations with ``Lint:`` and other non-PEP meta changes, such as updates to
> the Readme/Contributing Guide, issue/PR template, etc., with ``Meta:``.

Author: That was the convention I was following before, and I'm okay with that if others agree (provided we _do_ ensure that changes to specific PEPs are prefixed properly).

I don't think its necessary to be too prescriptive on this either (we have GitHub tags now anyway); this mostly was to avoid individual PEP editors coming up with their own cryptic, bespoke initialisms and confusing multiple other PEP editors, but that seems to be a pretty limited and specific problem that is better solved in other more appropriate ways.

```suggestion
```

developing at least a prototype reference implementation of your idea.

Contributing changes to existing PEPs
-------------------------------------

> In general, PEPs that are not marked as Draft or Active are usually considered

Reviewer 2: Remove one of "In general" and "usually". Come to think of it, probably remove both.

Author: I removed the latter; I wanted to hedge a bit here due to provisional PEPs, some PEPs that are de-facto specifications and updated from time to time, and other edge cases—though there has been a _general_ push to reduce that over time and move most of that elsewhere to more suitable platforms.

to be historical documents rather than documentation or living specifications.
To propose a substantially revised version of the original ideas,

> you should generally consider opening a new PEP instead.

Reviewer 2: Though not every change to something specified in a PEP will itself need a PEP to change.

Author: Yeah—if this sounds a bit out of place here, that's because it was originally part of the first bullet (Deferred/Withdrawn/Rejected), was more about do-overs for old PEPs. I realized it could apply, in somewhat modified form, to Final and Accepted PEPs too, so I moved it here.

In any case, to address this I completely re-wrote this paragraph to strengthn the conditions for a new PEP, add more alternatives, integrate the sentences better and add "Provisional" to "Active" and "Draft" above.

As such, make sure to consider the type and status of the PEP(s)
in question before proposing changes:

> * **Rejected**, **Deferred**, **Withdrawn** and **Superceded** PEPs should

Reviewer 2: Not sure about Deferred or Withdrawn. Deferred PEPs could see updates to bring them up to date; and a Withdrawn PEP could be prepared for a new shot.

Author: I was initially going to say something like this here (see above), but my general impression (which might be off base) was that for old PEPs that got enough out of step with modern Python to earn such a status (as the great majority of Deferred and Withdrawn PEPs are, since only PEPs >=several years old seem to have either status, aside from PEPs withdrawn for other reasons) it was usually recommended to start fresh with a new PEP rather than completely re-writing an old one. But that might not be an accurate impression, since I can't point to somewhere in PEP 1 where it says that (as opposed to for Rejected PEPs).

In any case, I've revised this bullet point to offer revival as a possibility.

  usually only be modified when necessary to correct technical issues with
  header processing or that prevent content from rendering as originally
  intended.

* **Final** PEPs are considered complete; while small fixes to clearly

>   unintentional errata in the document that meaningfully impair understanding

Reviewer 2: ```suggestion
  unintentional errors in the document that meaningfully impair understanding
```
Errata fix errors

Author: Gotcha, corrected

  may be considered, substantive changes to the content generally will not.

* **Accepted** PEPs may be updated to reflect minor details of the final
  implementation, but significant changes to the normative specification
  should be avoided.

* **Provisional** PEPs can be incrementally modified if needed, though
  major rewrites and compatibility breakage are discouraged unless absolutely
  necessary and any changes should be preceded by discussion and general
  consensus among the authors and the relevant community.

* **Active** PEPs can and should be updated whenever the information described
  changes. Substantive changes to governance PEPs should be reviewed by the
  Steering Council (by opening an
  `issue <https://github.com/python/steering-council/issues>`__ in the SC repo)
  and are subject to any further conditions stated in the PEP itself.

* **Draft** PEPs are freely open for discussion and proposed modification,
  at the discretion of the author. Copyediting, formatting and proofreading
  can be opened here as pull requests, while more substantive content
  changes should generally be first proposed on the PEP's discussion thread

>   listed in its ``Discussions-To`` header.

Reviewer 2: We should also mention that once the PEP has been submitted for pronouncement to the SC or delegate, further changes could delay the process.

Author: Good point; I revised this paragraph accordingly

developing at least a prototype reference implementation of your idea.

Contributing changes to existing PEPs
-------------------------------------

In general, PEPs that are not marked as Draft, Active or Provisional are
considered to be historical documents rather than living specifications
or documentation. Major changes to their core content usually require
a new PEP, while smaller modifications may or may not be appropriate,
depending on the PEP's status:

* **Rejected**, **Deferred**, **Withdrawn** and **Superceded** PEPs should
  usually only be modified when necessary to correct technical issues with
  header processing or that prevent content from rendering as originally
  intended. Occasionally, a Deferred or even Withdrawn PEP may be resurrected
  with major updates, but it is often better to just propose a new one.

* **Final** PEPs are considered complete; while small fixes to clearly
  unintentional errors in the document that meaningfully impair understanding
  may be considered, substantive changes to the content generally will not.

* **Accepted** PEPs may be updated to reflect minor details of the final

>   implementation, but significant changes to the normative specification

Reviewer 3: Is there a plain English synonym for "normative" that can be used? I had to look it up :)

Reviewer 2: We can just remove "normative".

Author: I just removed it; its an extra word and isn't critical to meaning here.

EDIT: Seems like I forgot to both actually push my changes from many hours ago, and submit this reply/resolve this comment. Oops, my fault.

  while more substantive content changes should generally be first proposed
  on the PEP's discussion thread listed in its ``Discussions-To`` header.

Also, while fixing spelling and formatting issues as part of more substantive
copyediting and proofreading of draft and active PEPs is encouraged,
we generally advise against PRs that simply mass-correct minor typos
that don't significantly impair meaning and understanding.

If you're unsure, in addition to the general guidance above,
we encourage you to reach out first before opening a PR here.
For example, you could contact the PEP author(s),
propose your idea in a discussion venue appropriate to the PEP (e.g.
`Typing-SIG <https://mail.python.org/archives/list/typing-sig@python.org/>`__
for static typing, or the `Packaging Discourse
<https://discuss.python.org/c/packaging/>`__ for packaging), and/or
open an `issue on this repo <https://github.com/python/peps/issues>`__.

Commit messages and PR titles
-----------------------------

When adding or modifying a PEP, please always include the PEP number in the
commit summary and pull request title.
For example, ``PEP NNN: <summary of changes>``.

> Likewise, prefix rendering infrastructure changes with ``Infra:``,

Reviewer 2: I think we can skip this sentence for brevity. It's not that important that the prefixes are consistent (though I agree with you that we should avoid obscure abbreviations like "PRS").

Author: We can, though I don't think it hurts to be provide some basic guidance here, since a lack of such was how we got cryptic ad-hoc initialisms like "PRS" in the first place, and its helpful to see at a glance the general high-level topic of a commit/PR without having to parse the full title/commit message.

Author: We could also drop the "Lint" prefix and just have "Infra" and "Meta", if that would simplify things.

Reviewer 3: "Lint", "Infra" and "Meta" are good.

I'm also confused about "PRS", what does it stand for? It's very difficult to search for :)

Author: Seems like opinions were mixed about specifying particular prefixes (@hugovk and me :+1: , @JelleZijlstra and @AA-Turner :+1: ), but us three plus Guido all found "PRS" very confusing, so I dropped this sentence for now on the condition that we all avoid bespoke initialisms like that going forward. Deal?

> I'm also confused about "PRS", what does it stand for? It's very difficult to search for :)

AFAIK @AA-Turner made it up, apparently (per his responses to Guido, Jelle, and me who all asked the same question). I believe it was intended to stand for "PEP Rendering System".
